lib/msal-core/src/Configuration.ts

```diff
@@ -0,0 +1,181 @@
+  cacheLocation?: CacheLocation;
+  storeAuthStateInCookie?: boolean;
+};
+
+/**
+ * Library Specific Options
+ *
+ * logger                       - Used to initialize the Logger object; TODO: Expand on logger details or link to the documentation on logger
+ * loadFrameTimeout             - maximum time the library should wait for a frame to load
+ * tokenRenewalOffsetSeconds    - sets the window of offset needed to renew the token before expiry
+ *
+ */
+export type SystemOptions = {
+  logger?: Logger;
+  loadFrameTimeout?: number;
+  tokenRenewalOffsetSeconds?: number;
+};
+
+/**
+ * App/Framework specific environment Support
+ *
+ * isAngular                - flag set to determine if it is Angular Framework. Used to broadcast tokens. TODO: detangle this dependency from core.
+ * unprotectedResources     - Array of URI's which are unprotected resources. MSAL will not attach a token to outgoing requests that have these URI. Defaults to 'null'.
+ * protectedResourceMap     - This is mapping of resources to scopes used by MSAL for automatically attaching access tokens in web API calls.
+ *                              A single access token is obtained for the resource. So you can map a specific resource path as follows:
+ *                              {"https://graph.microsoft.com/v1.0/me", ["user.read"]},
+ *                              or the app URL of the resource as: {"https://graph.microsoft.com/", ["user.read", "mail.send"]}.
+ *                              This is required for CORS calls.
+ *
+ */
+export type FrameworkOptions = {
+  isAngular?: boolean;
+  unprotectedResources?: Array<string>;
+  protectedResourceMap?: Map<string, Array<string>>;
+};
+
+/**
+ * Configuration Object
+ */
+export type Configuration = {
+  auth: AuthOptions,
+  cache?: CacheOptions,
+  system?: SystemOptions,
+  framework?: FrameworkOptions
+};
+
+/**
+ * Function to set the default options when not explicitly set
+ *
+ * @param TAuthOptions
+ * @param TCacheOptions
+ * @param TSystemOptions
+ * @param TFrameworkOptions
+ *
+ * @returns TConfiguration object
+ */
+
+// destructure with default settings
+export function buildConfiguration(
+  {
+    clientId = "",
+    authority = null,
+    validateAuthority = true,
+    redirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    postLogoutRedirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    state = "",
+    navigateToLoginRequestUrl = true
+  }: AuthOptions = { clientId },
+  {
+    cacheLocation = "sessionStorage",
+    storeAuthStateInCookie = false
+  }: CacheOptions = {},
+  {
+    logger = new Logger(null),
+    loadFrameTimeout = FRAME_TIMEOUT,
+    tokenRenewalOffsetSeconds = OFFSET
+  }: SystemOptions = {},
+  {
+    isAngular = false,
+    unprotectedResources = new Array<string>(),
+    protectedResourceMap = new Map<string, Array<string>>()
+  }: FrameworkOptions = {}
+): Configuration {
+  // restructure
+  let config: Configuration = {
```

Comment:
ill make the case one more time for why I do think destructing params here is harder to read and maybe a little duplicative 
 
```js
const DEFAULT_PARAMS = {
    auth: DEFAULT_AUTH_PARMAMS,
    cache: DEFAULT_CACHE_PARAMS,
    system: DEFAULT_SYSTEM_PARAMS,
    framework: DEAFULT_FRAMEWORK_PARAMS
};

...

export const buildConfig = config => ({ ...DEFAULT_CONFIG, config}); 
```

Reply:
The de-structuring allows me the pain of merging initiated params. It allows sparse param initialization and pick up the delta with defaults.The user can choose to:

- initialize only select params and have the remaining fill in default
- maintain the structure that would be a json may be eventually

I can write some test code and see if the above format works the same way. One thing that will be working out here for us is that even if the initialization style changes, we still can keep the base structure same. So I would like to keep this open for now and revisit it soon.

Reply:
This should do the same, essentially use default as a base and spread any passed in params over it.
Not a big deal now if you don't want to, it's a style thing and we can come back and clean up if we want,

Reply:
Tag this with samples, if we can get the same structure without extra code, I don't mind a different style. I need the fallback to defaults to be:

1) Readable
2) No extra checking of each parameter if it is initialized or not
3) No change in the substructures we created in the Config object

As long as the criteria are satisfied, I am fine. Resolving it here.